### PR TITLE
[RLlib] Add RLModule and MARLModule Serialize Deserialize methods

### PR DESCRIPTION
**Pull request by avnishn**
Signed-off-by: Avnish <avnishnarayan@gmail.com>

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/core/rl_module/rl_module.py [hunk 1]
```diff
@@ -302,6 +302,42 @@ def get_state(self) -> Mapping[str, Any]:
     def set_state(self, state_dict: Mapping[str, Any]) -> None:
         """Sets the state dict of the module."""
 
+    def serialize(self) -> Mapping[str, Any]:
```
Comment (kouroshHakha) [inline rllib/core/rl_module/rl_module.py:305]: I think this method should actually have a default implementation in here that is consistent with its deserialize counter-part.

File: rllib/core/rl_module/rl_module.py [hunk 2]
```diff
@@ -302,6 +302,42 @@ def get_state(self) -> Mapping[str, Any]:
...
+        NOTE: this state is typically obtained from `serialize()`.
+
+        NOTE: This method needs to be implemented in order to support
+            checkpointing and fault tolerance.
+
+        Returns:
+            A deserialized RLModule.
+        """
+        constructor = state["class"]
+        module = constructor(**state["kwargs"])
```
[the remark this answers is not in the document]
Comment (kouroshHakha) [reply rllib/core/rl_module/rl_module.py:337]: Also what if your kwargs are not serializable? They can be arbitrary python objects.  You need to implement serialization based on components that are absolutely serializable. This approach only works for when you have the constructor fully defined in the base-class, so it's clear how to serialize/deserialize the args and kwargs. One other candidate here is the information that is passed into `from_model_config()`, but it kinda defeats the purpose of that classmethod to make that information accessible inside the RLModule.

File: rllib/core/rl_module/marl_module.py [hunk 3]
```diff
@@ -358,6 +358,47 @@ def set_state(self, state_dict: Mapping[str, Any]) -> None:
...
+    def serialize(self) -> Mapping[str, Any]:
+        """Return the serialized state of the module.
+
+        NOTE: This method needs to be implemented in order to support
+        checkpointing and fault tolerance.
+
+        """
+        return {
+            "class": self.__class__,
+            "serialized_rl_modules": {
```
Comment (kouroshHakha) [inline rllib/core/rl_module/marl_module.py:370]: Call it `rl_modules`?

**Review by kouroshHakha — COMMENTED** [review]
I don't think this serialization is general-purpose enough. Please read my high-level concerns.

**Comment by kouroshHakha** [comment]
I think the solution here is to leave serialize/deserialize as abstract methods in RLModule since it doesn't have a settled constructor yet.

File: rllib/core/rl_module/rl_module.py (hunk 1 above)
Comment (avnishn) [reply rllib/core/rl_module/rl_module.py:317]: done

File: rllib/core/rl_module/rl_module.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (avnishn) [reply rllib/core/rl_module/rl_module.py]: resolved offline

File: rllib/core/rl_module/tf/tests/test_tf_rl_module.py [hunk 4]
```diff
@@ -95,6 +95,29 @@ def test_get_set_state(self):
         state2_after = module2.get_state()
         check(state, state2_after)
 
+    def test_serialize_deserialize(self):
```
Comment (kouroshHakha) [inline rllib/core/rl_module/tf/tests/test_tf_rl_module.py:98]: add a unit test that tests throwing an error for a non serializable construction of a module?

File: rllib/core/testing/tf/bc_module.py [hunk 5]
```diff
@@ -20,7 +20,9 @@ def __init__(
         hidden_dim: int,
         output_dim: int,
     ) -> None:
-        super().__init__()
+        super().__init__(
+            input_dim=input_dim, output_dim=output_dim, hidden_dim=hidden_dim
```
Comment (kouroshHakha) [inline rllib/core/testing/tf/bc_module.py:24]: for consistency either make both bc_modules (torch and tf) use kwargs or args.

**Review by kouroshHakha — APPROVED** [review]
Nice!! just a couple of minor comments. Thanks.

File: rllib/core/rl_module/tf/tests/test_tf_rl_module.py (hunk 4 above)
Comment (avnishn) [reply rllib/core/rl_module/tf/tests/test_tf_rl_module.py:99]: done

File: rllib/core/rl_module/tf/tests/test_tf_rl_module.py [hunk 6]
```diff
@@ -95,6 +96,44 @@ def test_get_set_state(self):
...
+        unpickleable_param = threading.Thread()
+
+        def bad_constructor():
+            return DiscreteBCTFModule(
+                input_dim=unpickleable_param,
+                hidden_dim=unpickleable_param,
+                output_dim=unpickleable_param,
+            )
+
+        self.assertRaisesRegex(
```
Comment (kouroshHakha) [inline rllib/core/rl_module/tf/tests/test_tf_rl_module.py:133]: Oh you are also checking the error message with a regex? I found it to make the unittests brittle for no reason when we make error messages more descriptive, just use `self.assertRaises(Exception, bad_constructor)` that's good enough.

**Review by kouroshHakha — APPROVED** [review]
Thanks, it looks good. One last comment about self.assertRaises vs. self.assertRaisesRegex.

File: rllib/core/rl_module/tf/tests/test_tf_rl_module.py (hunk 6 above)
Comment (avnishn) [reply rllib/core/rl_module/tf/tests/test_tf_rl_module.py]: resolved offline

File: rllib/core/rl_module/rl_module.py [hunk 7]
```diff
@@ -315,3 +362,29 @@ def as_multi_agent(self) -> "MultiAgentRLModule":
         from ray.rllib.core.rl_module.marl_module import MultiAgentRLModule
 
         return MultiAgentRLModule({DEFAULT_POLICY_ID: self})
+
+    @staticmethod
+    def __check_if_args_kwargs_serializable(args, kwargs):
```
Comment (gjoliver) [inline rllib/core/rl_module/rl_module.py:367]: dude, I feel like this should not be a static function on RLModule.
this is a very fundamental util that is worth living somewhere with a lot of unit tests of its own.

Comment (gjoliver) [reply rllib/core/rl_module/rl_module.py:367]: like, this has very little to do with RLModule.

Comment (avnishn) [reply rllib/core/rl_module/rl_module.py]: I moved this function to rllib/utils/serialization.py

**Pull request merged**